File: shuttle-api/src/ShuttleApi.Domain/Vehicles/VehicleServiceRecord.cs
using ShuttleApi.Domain.Common;

namespace ShuttleApi.Domain.Vehicles;

public sealed class VehicleServiceRecord : Entity<Guid>
{
    public Guid VehicleId { get; private set; }
    public ServiceCategory ServiceCategory { get; private set; }

    /// <summary>Only populated when ServiceCategory == FluidChange.</summary>
    public FluidType? FluidType { get; private set; }

    public string Title { get; private set; } = string.Empty;
    public string? Description { get; private set; }

    /// <summary>True = scheduled/planned maintenance; false = unplanned/emergency repair.</summary>
    public bool IsPlanned { get; private set; }

    public ServiceStatus ServiceStatus { get; private set; }
    public ServicePriority Priority { get; private set; }

    public DateTime? ScheduledDate { get; private set; }
    public DateTime? StartedDate { get; private set; }
    public DateTime? CompletedDate { get; private set; }

    public int? OdometerAtService { get; private set; }
    public decimal? EstimatedCostDollars { get; private set; }
    public decimal? ActualCostDollars { get; private set; }

    public string? ServiceProvider { get; private set; }
    public string? TechnicianName { get; private set; }
    public string? PartsNotes { get; private set; }
    public bool IsWarrantyWork { get; private set; }

    public DateTime? NextServiceDueDateUtc { get; private set; }
    public int? NextServiceDueOdometerKm { get; private set; }

    public DateTime CreatedAt { get; private set; }

    private VehicleServiceRecord() { }

    public static VehicleServiceRecord Create(
        Guid vehicleId,
        ServiceCategory serviceCategory,
        FluidType? fluidType,
        string title,
        string? description,
        bool isPlanned,
        ServiceStatus serviceStatus,
        ServicePriority priority,
        DateTime? scheduledDate,
        int? odometerAtService,
        decimal? estimatedCostDollars,
        string? serviceProvider,
        string? technicianName,
        string? partsNotes,
        bool isWarrantyWork,
        DateTime? nextServiceDueDateUtc,
        int? nextServiceDueOdometerKm)
    {
        return new VehicleServiceRecord
        {
            Id = Guid.NewGuid(),
            VehicleId = vehicleId,
            ServiceCategory = serviceCategory,
            FluidType = fluidType,
            Title = Guard.AgainstNullOrEmpty(title, nameof(title)),
            Description = description,
            IsPlanned = isPlanned,
            ServiceStatus = serviceStatus,
            Priority = priority,
            ScheduledDate = scheduledDate,
            OdometerAtService = odometerAtService,
            EstimatedCostDollars = estimatedCostDollars,
            ServiceProvider = serviceProvider,
            TechnicianName = technicianName,
            PartsNotes = partsNotes,
            IsWarrantyWork = isWarrantyWork,
            NextServiceDueDateUtc = nextServiceDueDateUtc,
            NextServiceDueOdometerKm = nextServiceDueOdometerKm,
            CreatedAt = DateTime.UtcNow
        };
    }

    public void Update(
        ServiceCategory serviceCategory,
        FluidType? fluidType,
        string title,
        string? description,
        bool isPlanned,
        ServiceStatus serviceStatus,
        ServicePriority priority,
        DateTime? scheduledDate,
        int? odometerAtService,
        decimal? estimatedCostDollars,
        string? serviceProvider,
        string? technicianName,
        string? partsNotes,
        bool isWarrantyWork,
        DateTime? nextServiceDueDateUtc,
        int? nextServiceDueOdometerKm)
    {
        ServiceCategory = serviceCategory;
        FluidType = fluidType;
        Title = Guard.AgainstNullOrEmpty(title, nameof(title));
        Description = description;
        IsPlanned = isPlanned;
        ServiceStatus = serviceStatus;
        Priority = priority;
        ScheduledDate = scheduledDate;
        OdometerAtService = odometerAtService;
        EstimatedCostDollars = estimatedCostDollars;
        ServiceProvider = serviceProvider;
        TechnicianName = technicianName;
        PartsNotes = partsNotes;
        IsWarrantyWork = isWarrantyWork;
        NextServiceDueDateUtc = nextServiceDueDateUtc;
        NextServiceDueOdometerKm = nextServiceDueOdometerKm;
    }

    /// <summary>Transitions the record to InProgress and records the start time.</summary>
    public void Start()
    {
        Guard.Against(ServiceStatus == ServiceStatus.Completed, "Cannot start a completed service record.");
        Guard.Against(ServiceStatus == ServiceStatus.Cancelled, "Cannot start a cancelled service record.");
        ServiceStatus = ServiceStatus.InProgress;
        StartedDate = DateTime.UtcNow;
    }

    /// <summary>Transitions the record to Completed.</summary>
    public void Complete(DateTime completedDate, decimal? actualCostDollars, int? odometerAtService)
    {
        Guard.Against(ServiceStatus == ServiceStatus.Cancelled, "Cannot complete a cancelled service record.");
        ServiceStatus = ServiceStatus.Completed;
        CompletedDate = completedDate;
        ActualCostDollars = actualCostDollars;
        if (odometerAtService.HasValue)
            OdometerAtService = odometerAtService;
    }
}

File: shuttle_tablet/lib/features/vehicles/domain/entities/vehicle_service_record.dart
import 'package:equatable/equatable.dart';

class VehicleServiceRecord extends Equatable {
  final String id;
  final String vehicleId;
  final String serviceCategory;
  final String? fluidType;
  final String title;
  final String? description;
  final bool isPlanned;
  final String serviceStatus;
  final String priority;
  final DateTime? scheduledDate;
  final DateTime? startedDate;
  final DateTime? completedDate;
  final int? odometerAtService;
  final double? estimatedCostDollars;
  final double? actualCostDollars;
  final String? serviceProvider;
  final String? technicianName;
  final String? partsNotes;
  final bool isWarrantyWork;
  final DateTime? nextServiceDueDateUtc;
  final int? nextServiceDueOdometerKm;
  final DateTime createdAt;

  const VehicleServiceRecord({
    required this.id,
    required this.vehicleId,
    required this.serviceCategory,
    this.fluidType,
    required this.title,
    this.description,
    required this.isPlanned,
    required this.serviceStatus,
    required this.priority,
    this.scheduledDate,
    this.startedDate,
    this.completedDate,
    this.odometerAtService,
    this.estimatedCostDollars,
    this.actualCostDollars,
    this.serviceProvider,
    this.technicianName,
    this.partsNotes,
    required this.isWarrantyWork,
    this.nextServiceDueDateUtc,
    this.nextServiceDueOdometerKm,
    required this.createdAt,
  });

  bool get isCompleted => serviceStatus.toLowerCase() == 'completed';
  bool get isOverdue =>
      isPlanned &&
      scheduledDate != null &&
      scheduledDate!.isBefore(DateTime.now()) &&
      (serviceStatus.toLowerCase() == 'scheduled' ||
          serviceStatus.toLowerCase() == 'deferred');

  @override
  List<Object?> get props => [
        id,
        vehicleId,
        serviceCategory,
        fluidType,
        title,
        description,
        isPlanned,
        serviceStatus,
        priority,
        scheduledDate,
        startedDate,
        completedDate,
        odometerAtService,
        estimatedCostDollars,
        actualCostDollars,
        serviceProvider,
        technicianName,
        partsNotes,
        isWarrantyWork,
        nextServiceDueDateUtc,
        nextServiceDueOdometerKm,
        createdAt,
      ];
}
